### PR TITLE
ext_proc: Support trailer callbacks

### PR DESCRIPTION
**Pull request by gbrail**
Commit Message: Support adding and modifying request and response trailers
in the ext_proc filter.

Additional Description:

Existing trailers will be sent to the processing
server if the processing mode is set to enable them. If the
processing mode is set to sent trailers, but there are no trailers
present, then empty trailers will be sent to the server for
modification.

However, trailers may only be added in the end of the data callback
in Envoy, which may come in before a previous gRPC reply returns.
Filters that need to be able to consistently add trailers where none
existed should enable trailer processing in the Envoy filter
configuration instead of relying on being able to turn it on
dynamically.

Risk Level: Low. Trailers only enabled if a service called by the filter
is configured to ask for them.

Testing: New integration and unit tests added.

Docs Changes: API docs updated in .proto files.

Release Notes:

When the processing mode is changed to SEND
for request or response trailers, a corresponding message will be
sent to the server, which can respond with trailer mutations as desired.

In addition, if trailer processing is enabled in the filter
configuration, then trailer messages will be sent to the server
even if trailers are not present. This makes it possible for the server
to add trailers where none exist.

Finally, at the moment Envoy only implements trailers for the HTTP/2
protocol. Nothing will happen if trailer processing is enabled and
Envoy is using HTTP/1 until Envoy implements trailers for HTTP/1.

Signed-off-by: Gregory Brail <gregbrail@google.com>

### CODE REVIEW

File: api/envoy/service/ext_proc/v3alpha/external_processor.proto
Comment (htuch) [inline api/envoy/service/ext_proc/v3alpha/external_processor.proto:77]: I just reread the `async_mode` field comments with a fresh pair of eyes and they are quite confusing. Can you try and clarify? I'm not sure here why `async_mode` basically replaces `response_required`.

File: api/envoy/service/ext_proc/v3alpha/external_processor.proto
Comment (htuch) [inline api/envoy/service/ext_proc/v3alpha/external_processor.proto:98]: Is there are any value in distinguishing between an Envoy client that sends an empty trailers frame and no trailers frame?

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:109]: Nit: gratuitous blank line

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:117]: Nit: `new_trailers`

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.cc:47]: Seems fragile if we have future reordering/additions to do >=.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.cc:71]: Why don't we track timeout for trailers?

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.h:29]: Can you remind me if, as with Envoy filter chains, it's possible to have both a headers and body message to the server in-flight at the same time?

**Review by htuch — CHANGES_REQUESTED** [review]
Exciting to see more of the implementation arriving :)
/wait

File: api/envoy/service/ext_proc/v3alpha/external_processor.proto
Comment (gbrail) [reply api/envoy/service/ext_proc/v3alpha/external_processor.proto:77]: All I'm doing in this PR is fixing the comment, since we renamed the flag from "response_required" to "async_mode" a long time ago while reviewing the original protocol. However, I made a mistake in the original PR. I just language to say "Unless async mode..." which is clearer IMO.

I remember a discussion on the original PR and we agreed that "async_mode" was better than "response_required" but TBH I thought that on this particular message, "response_required" is easier for a developer to understand.

File: api/envoy/service/ext_proc/v3alpha/external_processor.proto
Comment (gbrail) [reply api/envoy/service/ext_proc/v3alpha/external_processor.proto:98]: The only value IMO is in making it clearer to the server developer what to expect. I think it's easy to check whether the trailers frame will be empty, but if someone wants to write a filter that, for example, checksums the response body and adds the result in a trailer, then they need to be able to add trailers when they don't exist.

However, I could also see the need for a mode that says, basically, "only send me trailers if they already exist." But that's adding yet more logic here so I was hoping to keep it simpler for now.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:109]: Ack.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.cc:47]: I think you're referring to the >=, right? Let me review all the logic again and see if two or three bool flags might be easier to understand than one enum.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.cc:71]: We're just cancelling the timer that was set for the body response. In the call to filter_.sendTrailers() we'll re-enable it for the timers message.

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.h:29]: Not in the ext_proc protocol right now -- the filter goes to some trouble now to wait for the response to the headers message before sending the body message, and to wait for the body before trailers, and to do all that while assuming that the filter might be configured to send any combination of those three. I think that makes things simplest on the server, and it lets the server decide if it wants to see the body at all only after seeing the headers -- I think that's important, actually.

If we end up with a lot of servers that don't want to make this decision dynamically then there are more options that we could add, like pipelining the headers and body messages.

File: api/envoy/service/ext_proc/v3alpha/external_processor.proto
Comment (gbrail) [reply api/envoy/service/ext_proc/v3alpha/external_processor.proto]: I simplified the comments a bit so PTAL.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: I had to do all that ordering stuff to understand the state transitions and now that I figured that out we can get this done with a much simpler state mechanism, so PTAL the latest state of the code.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:107]: Non-actionable comment but I'm interested in your take; how adversarial are we treating remote ext_proc servers as being? One use case that might be interesting for `ext_proc` is allowing service chaining where trust relationships between infrastructure providers and tenants imply a threat model that doesn't fully trust the `ext_proc` server. Any thoughts on what would be required to make this fully robust?

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:117]: What happens if we bail early in a `return` below? It's a bit hard to follow if we're definitely going to send.

**Review by htuch — COMMENTED** [review]
LGTM modulo one small comment.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:119]: I updated the code to make the comments clearer, and renamed the "send_trailers" flag to make it easier to understand I hope. The logic is still correct -- we need to add the trailers here now because at least according to the comments in filter.h, we can only do it here, even if we're going to return right away and set the headers later.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:107]: Right now the relationship is pretty trusting -- ext_proc servers can set most headers, and they can manipulate the body and trailers too. There is an "ignore errors" setting which makes Envoy more resilient to a server failure, but without it a flaky server can definitely cause errors in the proxy as well. 

I think that if we don't trust servers, then the user should certainly set up ext_proc to ignore errors from servers, and we might want to consider a model in which a server is limited by the proxy as to what it is allowed to change in the headers, trailers, and body.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (htuch) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:119]: I guess I still find it a bit misleading having `setTrailersDelivered(true)` set here when they haven't yet been delivered. Is there a way to make this clearer in the state object?

Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:119]: What if instead of "completeBodyDelivered" and "trailersDelivered" flags in that state object, we rename them to "completeBodyAvailable" and "trailersAvailable." That would show that those flags are related to what happened in the filter state, and not what the actual filter has done so far. Does that make it clearer? If so I'll make the change.

Comment (htuch) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:119]: SGTM

Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: Done in the latest commit.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
